### ISSUE TITLE
Research perf improvements for queries with requireTotalCount

### ISSUE THREAD

**Issue by oliversturm** [oliversturm/devextreme-query-mongodb#31]
... or even any count. For some reason, even the most basic queries that require count seem to run slowly. This example runs fast:

```
skip=0&take=100&sort=[{%22selector%22:%22_id%22,%22desc%22:false}]&requireTotalCount=false
```

But when count is included, it gets slow with large document counts. Why? Doesn't MongoDB know how many documents there are in the collection? Surely a count query without a filter should always be fast? Apparently it's not, but the reasons behind this require research. (Note to self: I have not verified that the example demonstrates the issue as described.)

There are built-in functions to return document count and "estimated" document count:

https://docs.mongodb.com/manual/reference/method/db.collection.countDocuments/#db.collection.countDocuments
https://docs.mongodb.com/manual/reference/method/db.collection.estimatedDocumentCount/

**Comment by oliversturm** [comment]
Okay, tested this - a query that only retrieves the first 100 documents can be very fast (i.e. 1-3 millis). For 1 million documents, a pipeline that does nothing but a count (`aggregate([{$count:"count"}])`) already takes 419 millis (second test: 372).

**Comment by oliversturm** [comment]
`db.data.countDocuments({})` takes similar time - 424ms, 390ms, according to database log output - which is not surprising because the log shows that this command is translated into a very similar `aggregate` call as the one I run manually. For this reason I suspect that there wouldn't be any change with different criteria.

**Comment by oliversturm** [comment]
`db.data.estimatedDocumentCount()` seems to be faster, though there is no log output to measure it. However, this function can't accept a filter criteria and is therefore useless for the general purpose of the DevExtreme query.

**Comment by oliversturm** [comment]
`db.runCommand({ count: 'data' })` runs very fast. It is possible to specify a `query` parameter, and with an empty query the command is still very fast - this is perhaps an important find because it might speed up some cases, including the one in the original post above.

If the query parameter carries an actual criterion, the execution time is much the same as for any other variation above.

**Comment by oliversturm** [comment]
There might be some caveats with this method of counting: https://docs.mongodb.com/manual/reference/command/count/#dbcmd.count

However, it could be a good idea to use this method by default and offer a fallback - this may improve count behavior for filter-less queries, which is a relevant case because that's how any standard paged Data Grid setup would begin.

**Comment by oliversturm** [comment]
Turns out I can't use this as the default mechanism... for my full filter/search support I need to add nested fields (for expressions like `dateTimeField.Month`), and that's not possible if there's no pipeline. Even supporting search and filter at the same time isn't possible right now, because they use individual pipeline stages.

So... can only support this count feature as an option with caveats.

**Comment by oliversturm** [comment]
Fascinating. Now I was half done with an implementation using `db.collection.count()`, just to find that this particular function is deprecated now. Docs state:

> MongoDB drivers compatible with the 4.0 features deprecate their respective cursor and collection count() APIs in favor of new APIs for countDocuments() and estimatedDocumentCount(). For the specific API names for a given driver, see the driver documentation.

I'm not following. This function does something, according to my tests, that none of the other two functions do. So why is it deprecated?

**Comment by oliversturm** [comment]
The [docs for the top-level `count` function](https://docs.mongodb.com/manual/reference/command/count/#dbcmd.count) have the same deprecation warning. Excellent. WTF?

**Comment by oliversturm** [comment]
I can only imagine that the old functions are being deprecated due to the caveats mentioned in the docs: their counts may not always be accurate because they are based on metadata. I believe there are scenarios where this wouldn't matter, and the performance advantage I can gain by using these functions *without* filter criteria seem to be considerable. But that doesn't help if the functions go away in the future...

**Comment by oliversturm** [comment]
According to [this forum post](https://developer.mongodb.com/community/forums/t/about-golang-mongo-driver-countdocuments/8112/2), my suspicion is correct.

**Comment by oliversturm** [comment]
So here's the deal. [The branch "fix-31"](https://github.com/oliversturm/devextreme-query-mongodb/tree/fix-31) contains the code for a processing option `preferMetadataCount`, which activates use of `collection.count()` under the following conditions:

* The option is set to `true`, obviously
* A total count is required, i.e. either `requireTotalCount` or `totalSummary` are configured in loadOptions
* No `preProcessingPipeline` is configured (this feature is not public anyway)
* Just one element is contained in any filter pipeline, i.e. one search or one configured filter. Such a search or filter can not use any features like field postfixes, which result in the creation of nested fields.

The use of `collection.count()` means that count information is returned by checking collection metadata, instead of counting documents as part of a query pipeline. This saves significant time under the right circumstances, although it obviously comes with the caveats documented on the various MongoDB websites (i.e. potential inaccuracies with sharded databases etc).

Finally, the MongoDB driver outputs a deprecation warning when it detects execution of this function. It is possible that the feature goes away in the future - I can't judge how likely this is.

At this time I am undecided whether to publish the feature or not. If you think you want to use it exactly like it is, please let's hear your opinion in this thread.

**Comment by oliversturm** [comment]
Okay - based on email feedback I received I'll publish the feature. It doesn't do any harm, it's opt-in only - perhaps it helps in some cases. But no guarantees for the future, I may remove it again if the MongoDB driver status changes.

**Comment by oliversturm** [comment]
This feature is available in [version 2.0.13](https://github.com/oliversturm/devextreme-query-mongodb/releases/tag/v2.0.13) and [info has been added to the Wiki](https://github.com/oliversturm/devextreme-query-mongodb/wiki/QueryAPI#processing-option-prefermetadatacount).
